plugins/lang-plan-a/src/main/java/org/elasticsearch/plan/a/Metadata.java

```diff
@@ -0,0 +1,598 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plan.a;
+
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.elasticsearch.plan.a.Definition.Cast;
+import org.elasticsearch.plan.a.Definition.Type;
+import org.elasticsearch.plan.a.PlanAParser.ExpressionContext;
+import org.elasticsearch.plan.a.PlanAParser.PrecedenceContext;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Metadata is a wrapper for all the data that is collected by the {@link Analyzer}.  Each node in the ANTLR parse tree
+ * will have one of the types of metadata to store information used either in a different node by the analyzer
+ * or by the {@link Writer} during byte code generation.  Metadata also contains several objects passed into the
+ * {@link Analyzer} and {@link Writer} used during compilation including the {@link Definition}, the source code,
+ * the root of the ANTLR parse tree, and the {@link CompilerSettings}.
+ */
+class Metadata {
+
+    /**
+     * StatementMetadata is used to store metadata mostly about
+     * control flow for ANTLR nodes related to if/else, do, while, for, etc.
+     */
+    static class StatementMetadata {
+        /**
+         * The source variable is the ANTLR node used to generate this metadata.
+         */
+        final ParserRuleContext source;
+
+        /**
+         * The lastSource variable will be set to true when the final statement from the root ANTLR node is about
+         * to be visited.  This is used to determine whether or not the auto-return feature is allowed to be used,
+         * and if a null return value needs to be generated automatically since a return value is always required.
+         */
+        boolean lastSource = false;
+
+        /**
+         * The topLoop variable will be set to true when a loop node is initially visited.  This will not be
+         * propagated down the parse tree afterwards, though.  This is used to determine
+         * whether or not inLoop should be set further down the tree.  Note that inLoop alone is not enough
+         * information to determine whether we are in the last statement of a loop because we may inside of
```

Comment:
Attempted to clean this up a bit.  This probably needs more description in the Analyzer to be completely clear, but I'll hold off on that for later.